File: 78-subsets/subsets.py
class Solution(object):
    def subsets(self, nums):
        n = len(nums)
        ans = []
        temp = []
        def solve(idx, n):
            if idx>=n:
                ans.append(temp[:])
                return
            temp.append(nums[idx])
            solve(idx+1, n)
            temp.pop()
            solve(idx+1, n)
        
        solve(0,n)
        return ans

        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        

File: 78-subsets/subsets.java
class Solution {
    public void solve(int[] nums, int idx, List<List<Integer>> ans, HashSet<List<Integer>> st, List<Integer> temp){
        if(idx>=nums.length){
            if(!st.contains(temp)){
            ans.add(new ArrayList<>(temp));
            st.add(new ArrayList<>(temp));
            }
            return;
        }
        temp.add(nums[idx]);
        solve(nums, idx+1, ans, st, temp);
        temp.remove(temp.size() - 1);
        solve(nums, idx+1, ans, st, temp);
        return;
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        HashSet<List<Integer>> st = new HashSet<>();
        List<Integer> temp = new ArrayList<>(); 
        solve(nums,0,ans, st, temp);
        return ans;
    }
}